feat: Added TLS support for Redis Datasource

AnnaHariprasad5123 opened the pull request:
Fixes #26723  

**What's in this PR :**
- Added TlsConfiguration file for holding TLS properties.
- Added TlsConfiguration field to the DatasourceConfiguration file.
- Added Some constants in RedisErrorMessages file to show missing validation messages.
- Added SSL Redis URL to RedisTlsUtils file.
- Added some test cases in RedisPluginTest.

**For more clarity :** [Video about Redis connection with TLS enabled](https://drive.google.com/file/d/179iDhDSTctOW8AIDz0hsn5d_aYeFJA2Y/view?usp=sharing)

**Screenshots :**

**Before :** Unable to connect Redis via TLS
![image](https://github.com/user-attachments/assets/c90091e6-88e4-45b2-8333-0853ad467368)

**After :** Able to connect Redis via TLS with or without Client Authentication redis setup
![image](https://github.com/user-attachments/assets/c6478a06-c5db-404b-bc89-9287c4d81634)
![image](https://github.com/user-attachments/assets/6e55999b-6631-4560-b0f2-74da3955d5f2)
![image](https://github.com/user-attachments/assets/d7db94b1-48fa-4562-8de2-bb46b373d0e8)
![image](https://github.com/user-attachments/assets/0b317786-ae47-46d0-b10f-198de508df89)

**Added Test cases :**
- itShouldValidateDatasourceWithTlsEnabledAndMissingCACertificate
- itShouldValidateDatasourceWithTlsEnabledAndMissingClientCertificate
- itShouldValidateDatasourceWithTlsEnabledAndMissingKey
- itShouldPassValidationWithTlsEnabledAndValidCertificates
![image](https://github.com/user-attachments/assets/6637025c-86aa-4246-96c8-a15b33f5cf8a)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
	- Introduced TLS configuration support for Redis connections, enhancing security.
	- Added a new configuration section in the Redis plugin for TLS settings, including options for enabling TLS, verifying certificates, and client authentication.

- **Bug Fixes**
	- Improved error handling with specific messages for missing TLS configuration elements, such as client certificates and client keys.

- **Tests**
	- Expanded test coverage for TLS-related functionality, validating various scenarios and ensuring appropriate error handling.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

NilanshBansal commented:
Need to prioritise review for this PR. 
cc: @appsmithorg/query-js-pod @carinanfonseca

AnnaHariprasad5123 commented:
Hi Appsmith Team,

Could someone please review this PR when you have a chance? Your feedback would be greatly appreciated.

Thank you!

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -0,0 +1,131 @@
+        return jedisPool;
+    }
+
+    private static PrivateKey loadPrivateKey(String clientKey) throws Exception {
+        clientKey = clientKey
+                .replace("-----BEGIN PRIVATE KEY-----", "")
+                .replace("-----END PRIVATE KEY-----", "")
+                .replaceAll("\\s", "");
+
+        byte[] keyBytes = Base64.getDecoder().decode(clientKey);
+        PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keyBytes);
+        KeyFactory kf = KeyFactory.getInstance("RSA");
+        return kf.generatePrivate(spec);
+    }
```
cc: @AnnaHariprasad5123 please check this suggestion

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -0,0 +1,131 @@
+        }
+
+        // Initialize SSL context with appropriate managers
+        sslContext.init(
+                requiresClientAuth ? keyManagerFactory.getKeyManagers() : null, trustManagers, new SecureRandom());
```
cc: @AnnaHariprasad5123

NilanshBansal replied:
Please ignore this @AnnaHariprasad5123

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisURIUtils.java:
```diff
@@ -13,9 +13,17 @@ public class RedisURIUtils {
     public static URI getURI(DatasourceConfiguration datasourceConfiguration) throws URISyntaxException {
         StringBuilder builder = new StringBuilder();
-        builder.append(REDIS_SCHEME);
+
+        if (datasourceConfiguration.getTlsConfiguration() != null
+                && datasourceConfiguration.getTlsConfiguration().getTlsEnabled()) {
+            builder.append(REDIS_SSL_SCHEME);
+        } else {
+            builder.append(REDIS_SCHEME);
+        }
```
cc: @AnnaHariprasad5123

AnnaHariprasad5123 commented on app/server/appsmith-plugins/redisPlugin/src/main/resources/form.json:
```diff
@@ -58,6 +58,78 @@
+                "value": true
+              }
+            },
+            {
+              "label": "Requires TLS Client Authentication",
+              "configProperty": "datasourceConfiguration.tlsConfiguration.requiresClientAuth",
+              "controlType": "CHECKBOX",
+              "initialValue": false,
+              "hidden": {
+                "path": "datasourceConfiguration.tlsConfiguration.tlsEnabled",
+                "comparison": "NOT_EQUALS",
+                "value": true
+              }
+            },
+            {
+              "label": "Upload Client Certificate",
+              "configProperty": "datasourceConfiguration.tlsConfiguration.clientCertificateFile",
+              "controlType": "FILE_PICKER",
+              "encrypted": true,
+              "hidden": {
+                "path": "datasourceConfiguration.tlsConfiguration.requiresClientAuth",
+                "comparison": "EQUALS",
+                "value": false
+              }
+            },
+            {
+              "label": "Upload Client Key",
+              "configProperty": "datasourceConfiguration.tlsConfiguration.clientKeyFile",
+              "controlType": "FILE_PICKER",
+              "encrypted": true,
+              "hidden": {
+                "path": "datasourceConfiguration.tlsConfiguration.requiresClientAuth",
+                "comparison": "EQUALS",
+                "value": false
+              }
+            }
```
It is not working.

AnnaHariprasad5123 commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -0,0 +1,126 @@
+            trustManagerFactory.init(trustStore);
+            trustManagers = trustManagerFactory.getTrustManagers();
+        } else {
+            trustManagers = new TrustManager[] {
+                new X509TrustManager() {
+                    @Override
+                    public X509Certificate[] getAcceptedIssuers() {
+                        return new X509Certificate[0];
+                    }
+
+                    @Override
+                    public void checkClientTrusted(X509Certificate[] certs, String authType) {}
+
+                    @Override
+                    public void checkServerTrusted(X509Certificate[] certs, String authType) {}
+                }
+            };
+        }
```
It is user preference to verify the CA certificate or not.

AnnaHariprasad5123 commented on app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/TlsConfiguration.java:
```diff
@@ -0,0 +1,36 @@
+@AllArgsConstructor
+public class TlsConfiguration {
+
+    @JsonView({Views.Public.class, FromRequest.class})
+    Boolean tlsEnabled;
+
+    @JsonView({Views.Public.class, FromRequest.class})
+    Boolean verifyTlsCertificate;
```
I will make some changes to avoid Null pointer exception. Instead of adding default values.

AnnaHariprasad5123 commented:
Hi @NilanshBansal. I've implemented most of the changes suggested by Code Rabbit and marked them with a thumbs-up. For some changes, I’ve added comments, and I’ll add the rest shortly. Meanwhile, could you please check the functionality of Redis TLS

NilanshBansal commented:
@carinanfonseca can you please check [this](https://ce-37106.dp.appsmith.com) DP from the design perspective?
You can go to Datasources page, click on Redis datasource -> Enable TLS to view the changes. 
You can add dummy credentials to proceed for design testing.

carinanfonseca commented:
Hi @AnnaHariprasad5123, thanks for your contribution. Can you check and ensure the spacing on the datasources page matches the attached screenshot? When I checked it after sign-up and on the regular datasources form, the spacing seemed off, with extra margins. Thank you! <img width="1210" alt="Screenshot 2024-11-06 at 09 40 26" src="https://github.com/user-attachments/assets/5200093c-2a1f-4354-83fd-b8a89229a82d">

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -0,0 +1,131 @@
+            CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+
+            // Load client certificate
+            String clientCertContent =
+                    new String(tlsConfiguration.getClientCertificateFile().getDecodedContent(), StandardCharsets.UTF_8);
+            X509Certificate clientCert = (X509Certificate)
+                    certificateFactory.generateCertificate(new ByteArrayInputStream(clientCertContent.getBytes()));
+
+            // Load client private key
+            String clientKey =
+                    new String(tlsConfiguration.getClientKeyFile().getDecodedContent(), StandardCharsets.UTF_8);
```
@AnnaHariprasad5123 can you please add error handling here for exceptions that can come up from generateCertificate() etc. Adding a try-catch block should be good for handling exceptions for `CertificateException` and `IOException` type

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -0,0 +1,131 @@
+                    new String(tlsConfiguration.getCaCertificateFile().getDecodedContent(), StandardCharsets.UTF_8);
+
+            CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+            X509Certificate caCert = (X509Certificate)
```
Similarly please add a try-catch exception handling here

AnnaHariprasad5123 commented:
Hi @carinanfonseca, it seems that every field in each datasource has a top margin of 16px. Everything appears the same to me.
**Redis datasource when SignUp :**
![image](https://github.com/user-attachments/assets/960c5aa1-8b42-48ef-a5ab-70fa84da688a)
**Redis datasource when SignIn :**
![image](https://github.com/user-attachments/assets/f3b9350b-3867-4387-8331-89d34c1c81fd)
**Postgres datasource when SignUp:**
![image](https://github.com/user-attachments/assets/865bf345-96a3-4499-8dfc-3b3282b40cf9)
**Postgres datasource when SignIn :**
![image](https://github.com/user-attachments/assets/cda4a889-5077-411c-8518-f56729f47746)

carinanfonseca commented:
All good, @AnnaHariprasad5123! Thank you

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/plugins/RedisPlugin.java:
```diff
@@ -261,9 +263,16 @@ public Mono<JedisPool> datasourceCreate(DatasourceConfiguration datasourceConfig
-                        JedisPool jedisPool = new JedisPool(poolConfig, uri, timeout);
-                        log.debug(Thread.currentThread().getName() + ": Created Jedis pool.");
-                        return jedisPool;
+                        if (datasourceConfiguration.getTlsConfiguration() != null
+                                && !datasourceConfiguration
+                                        .getTlsConfiguration()
+                                        .getTlsEnabled()) {
+                            JedisPool jedisPool = new JedisPool(poolConfig, uri, timeout);
+                            log.debug(Thread.currentThread().getName() + ": Created Jedis pool.");
+                            return jedisPool;
+                        } else {
+                            return createJedisPoolWithTLS(poolConfig, uri, timeout, datasourceConfiguration);
+                        }
```
@AnnaHariprasad5123 can you please check this.

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -0,0 +1,146 @@
+                CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+
+                // Load client certificate
+                String clientCertContent = new String(
+                        tlsConfiguration.getClientCertificateFile().getDecodedContent(), StandardCharsets.UTF_8);
+                X509Certificate clientCert = (X509Certificate)
+                        certificateFactory.generateCertificate(new ByteArrayInputStream(clientCertContent.getBytes()));
+
+                // Load client private key
+                String clientKey =
+                        new String(tlsConfiguration.getClientKeyFile().getDecodedContent(), StandardCharsets.UTF_8);
+                PrivateKey privateKey = loadPrivateKey(clientKey);
```
@AnnaHariprasad5123 please check this

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -0,0 +1,146 @@
+        }
+    }
+
+    private static PrivateKey loadPrivateKey(String clientKey) throws Exception {
+        try {
+            clientKey = clientKey
+                    .replace("-----BEGIN PRIVATE KEY-----", "")
+                    .replace("-----END PRIVATE KEY-----", "")
+                    .replaceAll("\\s", "");
+
+            byte[] keyBytes = Base64.getDecoder().decode(clientKey);
+            PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keyBytes);
+            KeyFactory kf = KeyFactory.getInstance("RSA");
+            return kf.generatePrivate(spec);
+        } catch (Exception e) {
+            // Catch all other exceptions
+            log.error("Unexpected error while loading private key: {}", e.getMessage(), e);
+            throw e;
+        }
+    }
```
@AnnaHariprasad5123 please add this suggestion. This will check for EC algo keys as well and also clear the sensitive data from the memory.

NilanshBansal reviewed (changes requested):
Added comments

NilanshBansal commented:
Tested the above changes on DP. Other than the comments above rest LGTM.

AnnaHariprasad5123 commented on app/server/appsmith-plugins/redisPlugin/src/test/java/com/external/plugins/RedisPluginTest.java:
```diff
@@ -157,6 +171,63 @@ public void itShouldTestDatasource() {
+
+        assertTrue(pluginExecutor.validateDatasource(datasourceConfiguration).isEmpty());
+    }
+
```
```
@Test
    public void itShouldTestDatasourceWithTlsEnabledAndValidCertificates() {
        DatasourceConfiguration datasourceConfiguration = createDatasourceConfigurationWithTLSEnabled();
        Mono<DatasourceTestResult> datasourceTestResultMono = pluginExecutor.testDatasource(datasourceConfiguration);

        StepVerifier.create(datasourceTestResultMono)
            .assertNext(datasourceTestResult -> {
                assertNotNull(datasourceTestResult);
            })
            .verifyComplete();
    }
    ```
    If this is ok, I will push this.

NilanshBansal replied:
Yes @AnnaHariprasad5123 please push the changes. I will review.

AnnaHariprasad5123 commented:
Hi @NilanshBansal , Could you review the changes

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -50,16 +50,28 @@ public static JedisPool createJedisPoolWithTLS(
+                try {
+                    privateKey = loadPrivateKey(clientKeyBytes);
+                    // Use the privateKey object as needed
+                } finally {
```
@AnnaHariprasad5123 should we add a catch block also here which catches the CertificateException?

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisTLSManager.java:
```diff
@@ -108,8 +120,9 @@ public void checkServerTrusted(X509Certificate[] certs, String authType) {}
             }
 
             // Initialize SSL context with appropriate managers
+            SecureRandom secureRandom = SecureRandom.getInstance("SHA1PRNG");
```
instead of this I think you can directly use new SecureRandom(). The earlier code was correct. 
Please check my comment [here](https://github.com/appsmithorg/appsmith/pull/37106#discussion_r1846008484)

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/test/java/com/external/plugins/RedisPluginTest.java:
```diff
@@ -157,6 +217,75 @@ public void itShouldTestDatasource() {
+        assertTrue(pluginExecutor.validateDatasource(datasourceConfiguration).isEmpty());
+    }
+
+    @Test
+    public void itShouldTestDatasourceWithTlsEnabledAndValidCertificates() {
+        DatasourceConfiguration datasourceConfiguration = createDatasourceConfigurationWithTLSEnabled();
+        Mono<DatasourceTestResult> datasourceTestResultMono = pluginExecutor.testDatasource(datasourceConfiguration);
+
+        StepVerifier.create(datasourceTestResultMono)
+                .assertNext(datasourceTestResult -> {
+                    assertNotNull(datasourceTestResult);
+                })
+                .verifyComplete();
+    }
```
Not required as we are using mock certificates

AnnaHariprasad5123 commented:
Hi @NilanshBansal, Could you check now.

NilanshBansal commented:
/build-deploy-preview skip-tests=true recreate=true

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/main/java/com/external/utils/RedisURIUtils.java:
```diff
@@ -3,19 +3,33 @@
         StringBuilder builder = new StringBuilder();
-        builder.append(REDIS_SCHEME);
+
+        if (datasourceConfiguration.getTlsConfiguration() != null
```
Nit: @AnnaHariprasad5123 can you please add a check for datasourceConfiguration to be not null. 
if (datasourceConfiguration != null && datasourceConfiguration.getTlsConfiguration() != null
                && datasourceConfiguration.getTlsConfiguration().getTlsEnabled())

NilanshBansal commented on app/server/appsmith-plugins/redisPlugin/src/test/java/com/external/plugins/RedisPluginTest.java:
```diff
@@ -55,6 +66,23 @@ public static void setup() {
+        return new UploadedFile(fileName, base64Content);
+    }
+
+    private TlsConfiguration addTLSConfiguration(DatasourceConfiguration datasourceConfiguration) {
```
@AnnaHariprasad5123 parameter datasourceConfiguration is not getting consumed for the function addTLSConfiguration(), can you check and remove if it is not required.

AnnaHariprasad5123 commented:
![image](https://github.com/user-attachments/assets/c52aa7ea-85ee-4302-9f68-e311c1a0e456)

NilanshBansal commented:
/build-deploy-preview skip-tests=true

NilanshBansal commented:
@AnnaHariprasad5123 I am getting this error on using my TLS configuration with Redis. The same configuration was working earlier when I uploaded the CA certificate file.
<img width="1440" alt="image" src="https://github.com/user-attachments/assets/63e129f7-0755-4d5e-8a1b-029458bee93b">

AnnaHariprasad5123 commented:
Hi @NilanshBansal, 

Find the videos below
[In Local Setup](https://drive.google.com/file/d/1AVYovZVbKVs6iQqxxHv4Df7OPlZFYGh9/view?usp=sharing)
[In Deployed Setup](https://drive.google.com/file/d/1Jh9Lf962xMceiWg9GnSYoTR6FS0U1uQv/view?usp=sharing)
